### PR TITLE
Implementing moshi adapter and using parcelables with data class

### PR DESCRIPTION
**Pull request by tomaszpolanski**

### CODE REVIEW

File: app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+package com.futurice.freesound.network.api.model.mapping
+
+import com.futurice.freesound.network.api.model.GeoLocation
+import com.squareup.moshi.FromJson
+import com.squareup.moshi.JsonDataException
+
+class GeoLocationJsonAdapter {
+
+    @FromJson fun fromJson(json: String): GeoLocation {
+        val coordinateList = json.trim()
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt]:
```
       return json.trim()
                .split(' ')
                .mapNotNull { it.toDoubleOrNull() }
                .takeIf { (it.size == 2) }
                ?.let { GeoLocation(it[0], it[1]) } ?:
                throw JsonDataException("Unable to deserialize latitude/long values from: $json")
```

Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt]: Format differently if you like.

Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt]: Makes me wonder if Kotlin Options should have `.mapSome` and `.mapNone` to align with these.

File: app/src/main/java/com/futurice/freesound/network/api/model/GeoLocation.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+package com.futurice.freesound.network.api.model
+
+import android.os.Parcel
+import com.google.auto.value.AutoValue
+
+import android.os.Parcelable
+
+// Using https://github.com/nekocode/android-parcelable-intellij-plugin-kotlin for Parcelable
+data class GeoLocation(
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/network/api/model/GeoLocation.kt]: Should we consider this https://github.com/grandstaish/paperparcel

File: app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ */
+package com.futurice.freesound.network.api.model.mapping
+
+import com.futurice.freesound.network.api.model.GeoLocation
+import com.squareup.moshi.FromJson
+import com.squareup.moshi.JsonDataException
+
+class GeoLocationJsonAdapter {
+
+    @FromJson fun fromJson(json: String): GeoLocation {
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt]: This way of formatting annotations seems to be the general community consensus. Let use that approach generally for Kotlin.

File: app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        val coordinateList = json.trim()
+                .split(' ')
+                .mapNotNull { it.toDoubleOrNull() }
+        return if (coordinateList.size == 2)
+            GeoLocation(latitude = coordinateList[0], longitude = coordinateList[1])
+        else
+            throw JsonDataException("Unable to deserialize latitude/long values from: $json")
+
+    }
+}
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt]: Missing new line

File: app/src/main/java/com/futurice/freesound/network/api/model/SoundSearchResult.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.futurice.freesound.network.api.model
+
+data class SoundSearchResult (
+    val count: Int,
+    val next: String?,
```
Comment (tomaszpolanski) [inline app/src/main/java/com/futurice/freesound/network/api/model/SoundSearchResult.kt]: In that case we would need to write a special adapter for it, right?

File: app/src/main/java/com/futurice/freesound/network/api/model/GeoLocation.kt (hunk 2 above)
Comment (tomaszpolanski) [inline app/src/main/java/com/futurice/freesound/network/api/model/GeoLocation.kt]: We could, it is annoying to migrate to data classes but still write more code

File: app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt (hunk 3 above)
Comment (tomaszpolanski) [inline app/src/main/java/com/futurice/freesound/network/api/model/mapping/GeoLocationJsonAdapter.kt]: `@Test` as well?

File: app/src/main/java/com/futurice/freesound/network/api/model/SoundSearchResult.kt (hunk 5 above)
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/network/api/model/SoundSearchResult.kt]: I've already written that: https://github.com/peter-tackage/kotlin-options/blob/master/kotlin-options-moshi-adapter/src/main/kotlin/com/petertackage/kotlinoptions/moshi/MoshiJsonAdapters.kt

File: app/src/main/java/com/futurice/freesound/network/api/model/GeoLocation.kt (hunk 2 above)
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/network/api/model/GeoLocation.kt]: Different PR.

**Pull request merged**